backend/howtheyvote/models/vote.py

```diff
@@ -71,18 +79,134 @@ class ProcedureStage(Enum):
+class AmendmentAuthorOriginalText:
+    pass
+
+
+@dataclasses.dataclass
+class AmendmentAuthorMembers:
+    pass
+
+
+@dataclasses.dataclass
+class AmendmentAuthorOrally:
+    pass
+
+
+@dataclasses.dataclass
+class AmendmentAuthorRapporteur:
+    pass
+
+
+AmendmentAuthor = (
+    AmendmentAuthorGroup
+    | AmendmentAuthorCommittee
+    | AmendmentAuthorOriginalText
+    | AmendmentAuthorMembers
+    | AmendmentAuthorOrally
+    | AmendmentAuthorRapporteur
+)
+
+AMENDMENT_AUTHOR_TYPE_TO_CLASS: dict[AmendmentAuthorType, type[AmendmentAuthor]] = {
+    AmendmentAuthorType.GROUP: AmendmentAuthorGroup,
+    AmendmentAuthorType.COMMITTEE: AmendmentAuthorCommittee,
+    AmendmentAuthorType.ORIGINAL_TEXT: AmendmentAuthorOriginalText,
+    AmendmentAuthorType.MEMBERS: AmendmentAuthorMembers,
+    AmendmentAuthorType.ORALLY: AmendmentAuthorOrally,
+    AmendmentAuthorType.RAPPORTEUR: AmendmentAuthorRapporteur,
+}
+
+AMENDMENT_AUTHOR_CLASS_TO_TYPE = {v: k for k, v in AMENDMENT_AUTHOR_TYPE_TO_CLASS.items()}
+
+
+def serialize_amendment_author(author: AmendmentAuthor | None) -> dict[str, Any] | None:
+    if not author:
+        return None
+
+    if isinstance(author, AmendmentAuthorGroup):
+        return {
+            "type": AmendmentAuthorType.GROUP,
+            "group": author.group.code if author.group else None,
+        }
+
+    if isinstance(author, AmendmentAuthorCommittee):
+        return {
+            "type": AmendmentAuthorType.COMMITTEE,
+            "committee": author.committee.code if author.committee else None,
+        }
 
+    return {"type": AMENDMENT_AUTHOR_CLASS_TO_TYPE[author.__class__]}
 
-class SerializedMemberVote(TypedDict):
+
+def deserialize_amendment_author(author: dict[str, Any] | None) -> AmendmentAuthor | None:
+    if not author:
+        return None
+
+    type_ = AmendmentAuthorType[author["type"]]
+
+    if type_ == AmendmentAuthorType.GROUP:
+        return AmendmentAuthorGroup(group=Group[author["group"]] if author["group"] else None)
+
+    if type_ == AmendmentAuthorType.COMMITTEE:
+        return AmendmentAuthorCommittee(
+            committee=Committee[author["committee"]] if author["committee"] else None,
+        )
+
+    author_class = AMENDMENT_AUTHOR_TYPE_TO_CLASS[type_]
+
+    # Typing this correctly would require a lot of boilerplate. Skipping that
+    # given that this is very isolated
+    return author_class()  # type: ignore
+
+
+class SAAmendmentAuthorType(TypeDecorator[AmendmentAuthor]):
+    impl = sa.JSON
+    cache_ok = True
+
+    def process_bind_param(
+        self, value: AmendmentAuthor | None, dialect: Dialect
+    ) -> dict[str, Any] | None:
+        if value is None:
+            return None
+
+        return serialize_amendment_author(value)
+
+    def process_result_value(
+        self, value: dict[str, Any] | None, dialect: Dialect
+    ) -> AmendmentAuthor | None:
+        if value is None:
+            return None
+
+        return deserialize_amendment_author(value)
```

Comment:
This way of modeling amendment authors is quite verbose, but I didn’t find a simpler way that’s still type safe, i.e. I want this to work without manual type casting or type errors:

```py
for author in vote.amendment_authors:
  if author.type == AmendmentAuthorType.GROUP:
    print(author.group.label)
  # ...
```